flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/remote/RemoteStorageScanner.java

```diff
@@ -0,0 +1,241 @@
+        this.baseRemoteStoragePath = baseRemoteStoragePath;
+        this.requiredSegmentIds = new ConcurrentHashMap<>();
+        this.scannedMaxSegmentIds = new ConcurrentHashMap<>();
+        this.scanStrategy = new ScanStrategy(INITIAL_SCAN_INTERVAL_MS, MAX_SCAN_INTERVAL_MS);
+        try {
+            this.remoteFileSystem = new Path(baseRemoteStoragePath).getFileSystem();
+        } catch (IOException e) {
+            ExceptionUtils.rethrow(
+                    e, "Failed to initialize file system on the path: " + baseRemoteStoragePath);
+        }
+    }
+
+    /** Start the executor. */
+    public void start() {
+        scannerExecutor.schedule(
+                this, scanStrategy.getInterval(attemptNumber), TimeUnit.MILLISECONDS);
+    }
+
+    /**
+     * Watch the segment for a specific subpartition in the {@link RemoteStorageScanner}.
+     *
+     * <p>If a segment with a larger or equal id already exists, the current segment won't be
+     * watched.
+     *
+     * <p>If a segment with a smaller segment id is still being watched, the current segment will
+     * replace it because the smaller segment should have been consumed. This method ensures that
+     * only one segment file can be watched for each subpartition.
+     *
+     * @param partitionId is the id of partition.
+     * @param subpartitionId is the id of subpartition.
+     * @param segmentId is the id of segment.
+     */
+    public void watchSegment(
+            TieredStoragePartitionId partitionId,
+            TieredStorageSubpartitionId subpartitionId,
+            int segmentId) {
+        Tuple2<TieredStoragePartitionId, TieredStorageSubpartitionId> key =
+                Tuple2.of(partitionId, subpartitionId);
+        if (scannedMaxSegmentIds.getOrDefault(key, -1) >= segmentId) {
+            return;
+        }
+        requiredSegmentIds.put(key, segmentId);
+    }
+
+    /** Close the executor. */
+    public void close() {
+        scannerExecutor.shutdownNow();
+    }
+
+    /** Iterate the watched segment ids and check related file status. */
+    @Override
+    public void run() {
+        Iterator<Map.Entry<Tuple2<TieredStoragePartitionId, TieredStorageSubpartitionId>, Integer>>
+                iterator = requiredSegmentIds.entrySet().iterator();
+        boolean scanned = false;
+        while (iterator.hasNext()) {
+            Map.Entry<Tuple2<TieredStoragePartitionId, TieredStorageSubpartitionId>, Integer> ids =
+                    iterator.next();
+            TieredStoragePartitionId partitionId = ids.getKey().f0;
+            TieredStorageSubpartitionId subpartitionId = ids.getKey().f1;
+            int requiredSegmentId = ids.getValue();
+            int maxSegmentId = scannedMaxSegmentIds.getOrDefault(ids.getKey(), -1);
+            if (maxSegmentId >= requiredSegmentId) {
+                scanned = true;
+                iterator.remove();
+                notifier.notifyAvailableAndPriority(partitionId, subpartitionId, false);
```

Comment:
This doesn't seem right. A required segment may not exist in remote tier at all. `maxSegmentId >= requiredSegmentId` does not necessarily mean the segment exist.